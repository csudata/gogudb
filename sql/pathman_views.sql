/*
 * -------------------------------------------
 *  NOTE: This test behaves differenly on 9.5
 * -------------------------------------------
 */

\set VERBOSITY terse

SET search_path = 'public','_gogu';
CREATE EXTENSION gogudb;
CREATE SCHEMA views;



/* create a partitioned table */
create table views._abc(id int4 not null);
select create_hash_partitions('views._abc', 'id', 10);
insert into views._abc select generate_series(1, 100);


/* create a facade view */
create view views.abc as select * from views._abc;

create or replace function views.disable_modification()
returns trigger as
$$
BEGIN
  RAISE EXCEPTION '%', TG_OP;
  RETURN NULL;
END;
$$
language 'plpgsql';

create trigger abc_mod_tr
instead of insert or update or delete
on views.abc for each row
execute procedure views.disable_modification();


/* Test SELECT */
explain (costs off) select * from views.abc;
explain (costs off) select * from views.abc where id = 1;
explain (costs off) select * from views.abc where id = 1 for update;
select * from views.abc where id = 1 for update;
select count (*) from views.abc;


/* Test INSERT */
explain (costs off) insert into views.abc values (1);
insert into views.abc values (1);


/* Test UPDATE */
explain (costs off) update views.abc set id = 2 where id = 1 or id = 2;
update views.abc set id = 2 where id = 1 or id = 2;


/* Test DELETE */
explain (costs off) delete from views.abc where id = 1 or id = 2;
delete from views.abc where id = 1 or id = 2;



DROP SCHEMA views CASCADE;
DROP EXTENSION gogudb;

