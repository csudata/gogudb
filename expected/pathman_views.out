/*
 * -------------------------------------------
 *  NOTE: This test behaves differenly on 9.5
 * -------------------------------------------
 */
\set VERBOSITY terse
SET search_path = 'public','_gogu';
CREATE EXTENSION gogudb;
CREATE SCHEMA views;
/* create a partitioned table */
create table views._abc(id int4 not null);
select create_hash_partitions('views._abc', 'id', 10);
 create_hash_partitions 
------------------------
                     10
(1 row)

insert into views._abc select generate_series(1, 100);
/* create a facade view */
create view views.abc as select * from views._abc;
create or replace function views.disable_modification()
returns trigger as
$$
BEGIN
  RAISE EXCEPTION '%', TG_OP;
  RETURN NULL;
END;
$$
language 'plpgsql';
create trigger abc_mod_tr
instead of insert or update or delete
on views.abc for each row
execute procedure views.disable_modification();
/* Test SELECT */
explain (costs off) select * from views.abc;
           QUERY PLAN            
---------------------------------
 Append
   ->  Seq Scan on _views_0__abc
   ->  Seq Scan on _views_1__abc
   ->  Seq Scan on _views_2__abc
   ->  Seq Scan on _views_3__abc
   ->  Seq Scan on _views_4__abc
   ->  Seq Scan on _views_5__abc
   ->  Seq Scan on _views_6__abc
   ->  Seq Scan on _views_7__abc
   ->  Seq Scan on _views_8__abc
   ->  Seq Scan on _views_9__abc
(11 rows)

explain (costs off) select * from views.abc where id = 1;
           QUERY PLAN            
---------------------------------
 Append
   ->  Seq Scan on _views_5__abc
         Filter: (id = 1)
(3 rows)

explain (costs off) select * from views.abc where id = 1 for update;
              QUERY PLAN               
---------------------------------------
 LockRows
   ->  Append
         ->  Seq Scan on _views_5__abc
               Filter: (id = 1)
(4 rows)

select * from views.abc where id = 1 for update;
 id 
----
  1
(1 row)

select count (*) from views.abc;
 count 
-------
   100
(1 row)

/* Test INSERT */
explain (costs off) insert into views.abc values (1);
  QUERY PLAN   
---------------
 Insert on abc
   ->  Result
(2 rows)

insert into views.abc values (1);
ERROR:  INSERT
/* Test UPDATE */
explain (costs off) update views.abc set id = 2 where id = 1 or id = 2;
                 QUERY PLAN                  
---------------------------------------------
 Update on abc
   ->  Result
         ->  Append
               ->  Seq Scan on _views_5__abc
                     Filter: (id = 1)
               ->  Seq Scan on _views_9__abc
                     Filter: (id = 2)
(7 rows)

update views.abc set id = 2 where id = 1 or id = 2;
ERROR:  UPDATE
/* Test DELETE */
explain (costs off) delete from views.abc where id = 1 or id = 2;
                 QUERY PLAN                  
---------------------------------------------
 Delete on abc
   ->  Result
         ->  Append
               ->  Seq Scan on _views_5__abc
                     Filter: (id = 1)
               ->  Seq Scan on _views_9__abc
                     Filter: (id = 2)
(7 rows)

delete from views.abc where id = 1 or id = 2;
ERROR:  DELETE
DROP SCHEMA views CASCADE;
NOTICE:  drop cascades to 13 other objects
DROP EXTENSION gogudb;
